### ISSUE TITLE
Bug in tf.image.random_contrast() ?

### ISSUE THREAD

**Issue by Hvass-Labs** [tensorflow/tensorflow#3816]
TensorFlow v. 0.10.0rc0 CPU on Linux.

In this file for the CIFAR-10 example:

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/models/image/cifar10/cifar10_input.py

Lines 178-179 are:

    distorted_image = tf.image.random_contrast(distorted_image, lower=0.2, upper=1.8)

But this gives the following result. Is this the intended behaviour of this function? Perhaps this should be explained better in the doc-string?

There is a similar problem with `tf.image.random_brightness()` and `adjust_brightness()` and its use in `cifar10_input.py`

![random_contrast](https://cloud.githubusercontent.com/assets/13588114/17666469/889ffc9e-6300-11e6-8163-9d7e23ca0f75.png)

**Comment by sherrym** [comment]
What do you expect the output to be?

**Comment by Hvass-Labs** [comment]
I suppose the output should be like the 3 images on the right-hand side? The other images are weirdly distorted. It seems to overflow or something when `tf.image.random_contrast()` is called with `upper > 1.0` and the cifar10 example calls it with `upper=1.8`

I should mention that I call these functions in my own code where the pixels are floats between zero and one. Perhaps that is the problem. I don't know because it's not clear from the doc-strings what the functions assume of the input. Please clarify the doc-strings in `tf.image`. Perhaps an `assert upper <= 1.0` would also be a good idea in the `random_contrast()` function, and similarly for the other functions in `tf.image`.

**Comment by aselle** [comment]
These do not necessarily look wrong to me. You are asking for contrast. The contrast formula outputs for each channel c at each pixel i,j
mean =sum(x(i,j,c),i,j)/(N*M)
x(i,j,c) = contrast * (x(i,j,c)-mean) + mean
This formula can produce values that are greater than 1.  This contrast formula is more general histogram equalization than what a paint program does. A paint program turns contrast into a pointwise operation typically by assuming a 0.5 mean on every channel. So the contrast behavior looks like what I would expect.  To see if you are getting a proper contrast adjustment, try doing contrast and then plotting a histogram of each channel before and after the contrast adjustment.

Try doing maximum contrast on a color photo in gimp, you'll see the same extreme pixels. The standard contrast there is limited to 127 (which is because they go [0,255] by default). That is equivalent to 0.5 contrast in the [0,1] image scale. So cifar is doing way more, so the extreme pixels will be even more extreme.

**Comment by Hvass-Labs** [comment]
Thanks for the comments.

I suppose there are two issues when the pixel values overflow beyond the [0, 1] range.

(1) Whether the classification performance of the network is affected. I think this might be the case because this was how I initially discovered the overflow, because my neural network performed worse. Although it's been several weeks now, so I can't remember the details.

(2) Plotting the images with pixels that have overflown shows weird pixels like in the images above.

A simple solution is to finalize the random distortions with two lines of code that limit the pixel values to [0, 1], like so:

    image = tf.image.random_contrast(image, lower=0.2, upper=1.8)

    # Do various other distortions ...

    # Limit pixel values to [0, 1]
    image = tf.minimum(image, 1.0)
    image = tf.maximum(image, 0.0)

Example with this limiting when doing various random distortions:

![limited](https://cloud.githubusercontent.com/assets/13588114/18047760/c1ad0070-6dde-11e6-8367-2990eef537c6.png)


Without the limiting:

![overflow](https://cloud.githubusercontent.com/assets/13588114/18047773/c98298aa-6dde-11e6-94ec-af1c5c5401a5.png)

Note: To get the same random values in these two sets of images, I first set the random seed using `tf.set_random_seed(42)`, but the two right-most images in the top-row have different random crops. I don't know why.

**Comment by aselle** [comment]
Yes. This is what I expected. I would say in general it is reasonable for float images to exceed 1. High dynamic range (HDR) images often will have this being the case. HDR images are not commonly used in ML. HDR images have to be tone-mapped (constrained) to LDR displays through some process. Contrast, brightness are simple forms of tone mapping, but there are many more complicated algorithms that can be used to preserve things like neighboring (relative contrast). 

The upshot is that it does not seem appropriate to clamp the values inside of contrast functions. If a user needs or wants to do this for training performance, using tf.minimum and tf.maximum can always be added.

**Comment by Hvass-Labs** [comment]
Thanks again for the comments.

I think it would be wise to add a note about this in the docs for the image functions; that their output might exceed the input range, and that an easy solution is to limit the output after a sequence of image operations. I imagine that most people don't have deep insights about these image algorithms and might be surprised about overflows if it's not in the docs.

Closed.

**Comment by chrisrapson** [comment]
Since my un-augmented test images will only have pixel values in the range from 0-1, I decided to restrict my augmented images to the same range. Instead of the `tf.minimum(image, 1.0)` suggestion from before, I preferred to scale the pixel values, rather than just cap them at 0 or 1. I also take the min/max over all channels, to minimise colour distortion.

It's just these 2 lines of code:
```
image -= tf.reduce_min(image)
image = image / tf.reduce_max(image)
```

These lines can be run once after all augmentation is complete. That is more efficient than adjusting the range after each operation, so I agree that the `tf.image.X` functions should not limit the range of their outputs. i.e. they should be left as they are.
